### PR TITLE
feat: commit to validator bitmap

### PR DESCRIPTION
**Pull request by ratankaliani**
Commit to validator bitmap for slashing equivocation. cc @evan-forbes

Update `README` with documentation for deploying the contract.

Get nonce from eth_getTransactionCount.

### CODE REVIEW

**Review by evan-forbes — APPROVED** [review]
[optional nit]
should this be a `feat!`?

LGTM!!

would be interested in a review from @rach-id as well

**Comment by ratankaliani** [comment]
@evan-forbes Don't think it should be a feature, support slashing equivocation should be on by default.

**Review by rach-id — COMMENTED** [review]
trying to run `cargo build` on this branch and I get:
``` 
   Compiling env_logger v0.11.3
   Compiling dotenv v0.15.0
   Compiling serde_cbor v0.11.2
   Compiling clap v4.5.7
   Compiling ethers-contract-derive v2.0.14
   Compiling sp1-recursion-program v0.1.0 (https://github.com/succinctlabs/sp1.git?rev=v1.0.5-testnet#19ef89a4)
The following warnings were emitted during compilation:

warning: blobstream-script@0.1.0: blobstream-program built at 2024-06-24 12:03:31

error: failed to run custom build command for `blobstream-script v0.1.0 (/Users/midnight/Tmp/sp1-blobstream/script)`

Caused by:
  process didn't exit successfully: `/Users/midnight/Tmp/sp1-blobstream/target/debug/build/blobstream-script-f34327e4a0958ae8/build-script-build` (exit status: 101)
  --- stdout
  path: "../program"
  cargo::rerun-if-changed=../program/src
  cargo::rerun-if-changed=../program/Cargo.toml
  cargo::rerun-if-changed=../program/Cargo.lock
  cargo:warning=blobstream-program built at 2024-06-24 12:03:31

  --- stderr
  [sp1] error: no such command: `prove`
  [sp1] 
  [sp1]         Did you mean `remove`?
  [sp1] 
  [sp1]         View all installed commands with `cargo --list`
  [sp1]         Find a package to install `prove` with `cargo search cargo-prove`
  thread 'main' panicked at /Users/midnight/.cargo/git/checkouts/sp1-20c98843a1ffc860/a8021d9/helper/src/lib.rs:47:9:
  Failed to build `blobstream-program`.
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
warning: build failed, waiting for other jobs to finish...
```

Also when running: `TENDERMINT_RPC_URL=https://rpc.lunaroasis.net/ cargo run --bin genesis --release`: 

```
   Compiling half v1.8.3
   Compiling serde_cbor v0.11.2
   Compiling env_logger v0.11.3
   Compiling dotenv v0.15.0
   Compiling tendermint-light-client-verifier v0.35.0
   Compiling primitives v0.1.0 (/Users/midnight/Tmp/sp1-blobstream/primitives)
The following warnings were emitted during compilation:

warning: blobstream-script@0.1.0: blobstream-program built at 2024-06-24 12:05:38

error: failed to run custom build command for `blobstream-script v0.1.0 (/Users/midnight/Tmp/sp1-blobstream/script)`

Caused by:
  process didn't exit successfully: `/Users/midnight/Tmp/sp1-blobstream/target/release/build/blobstream-script-87847c8a0557bcaa/build-script-build` (exit status: 101)
  --- stdout
  path: "../program"
  cargo::rerun-if-changed=../program/src
  cargo::rerun-if-changed=../program/Cargo.toml
  cargo::rerun-if-changed=../program/Cargo.lock
  cargo:warning=blobstream-program built at 2024-06-24 12:05:38

  --- stderr
  [sp1] error: no such command: `prove`
  [sp1] 
  [sp1]         Did you mean `remove`?
  [sp1] 
  [sp1]         View all installed commands with `cargo --list`
  [sp1]         Find a package to install `prove` with `cargo search cargo-prove`
  thread 'main' panicked at /Users/midnight/.cargo/git/checkouts/sp1-20c98843a1ffc860/a8021d9/helper/src/lib.rs:47:9:
  Failed to build `blobstream-program`.
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
warning: build failed, waiting for other jobs to finish...
```

File: program/src/main.rs [hunk 1]
```diff
@@ -97,13 +76,33 @@ fn main() {
...
+    for i in 0..trusted_light_block.signed_header.commit.signatures.len() {
+        if let CommitSig::BlockIdFlagCommit {
+            validator_address: _,
+            timestamp: _,
+            signature: _,
+        } = &trusted_light_block.signed_header.commit.signatures[i]
+        {
+            validator_bitmap[i] = true;
+        }
+    }
```
Comment (rach-id) [inline program/src/main.rs:93]: If I understand correctly, this is setting the bitmap for the trusted header. This is not enough to mitigate the discussed issue.

The bitmap should reflect the intersection of the validators that signed all the blocks in the range.

So, for example, if we have 6 validators, and we have:
- block 1: signed by 1,2,3,4,5
- block 2: signed by 3,4,5,6
- block 3: signed by 1,2,3,4,5
- block 4: signed by 4,5,6

Then, the bitmap should be: `[false, false, false, true, true, false]`

**Review by rach-id — COMMENTED** [review]
🔥 🚢  Amazing stuff. 
Left some comments and a discussion for the bitmap thingy. Also, would love to be able to build the project (check comments above for build errors).

**Comment by ratankaliani** [comment]
@rach-id Can you install the SP1 toolchain on your machine? The issue that you're running into is that you don't have the cargo prove (SP1) CLI: https://docs.succinct.xyz/getting-started/install.html#option-1-prebuilt-binaries-recommended

File: primitives/src/lib.rs [hunk 2]
```diff
@@ -1 +1,44 @@
...
+        if value {
+            let word_index = index / 64;
+            let bit_index = index % 64;
+            if word_index < 4 {
+                result[word_index] |= 1 << bit_index;
+            }
+        }
+    }
+    U256::from_limbs(result)
+}
```
[the remark this answers is not in the document]
Comment (ratankaliani) [reply primitives/src/lib.rs:44]: U256 from alloy is composed of 4 u64's, it's an alias for the ruint type.

[the remark this answers is not in the document]
Comment (rach-id) [reply primitives/src/lib.rs:44]: but why use it and not use the above code I shared?

[the remark this answers is not in the document]
Comment (ratankaliani) [reply primitives/src/lib.rs:44]: Misread the comment, changed to the above code!

**Review by rach-id — APPROVED** [review]
utAck 🔥

**Pull request merged**